src/builder.rs

```diff
@@ -372,98 +426,128 @@ impl Builder {
-        // Pair up the spends and recipients, extending with dummy values as necessary.
-        let pre_actions: Vec<_> = {
-            let num_spends = self.spends.len();
-            let num_recipients = self.recipients.len();
-            let num_actions = [num_spends, num_recipients, MIN_ACTIONS]
-                .iter()
-                .max()
-                .cloned()
-                .unwrap();
-
-            self.spends.extend(
-                iter::repeat_with(|| SpendInfo::dummy(&mut rng)).take(num_actions - num_spends),
-            );
-            self.recipients.extend(
-                iter::repeat_with(|| RecipientInfo::dummy(&mut rng))
-                    .take(num_actions - num_recipients),
-            );
-
-            // Shuffle the spends and recipients, so that learning the position of a
-            // specific spent note or output note doesn't reveal anything on its own about
-            // the meaning of that note in the transaction context.
-            self.spends.shuffle(&mut rng);
-            self.recipients.shuffle(&mut rng);
-
-            self.spends
-                .into_iter()
-                .zip(self.recipients.into_iter())
-                .map(|(spend, recipient)| ActionInfo::new(spend, recipient, &mut rng))
-                .collect()
-        };
+        self,
+        rng: impl RngCore,
+    ) -> Result<Option<Bundle<InProgress<Unproven, Unauthorized>, V>>, BuildError> {
+        bundle(rng, self.spends, self.outputs, self.bundle_type)
+    }
+}
 
-        // Move some things out of self that we will need.
-        let flags = self.flags;
-        let anchor = self.anchor;
+/// Builds a bundle containing the given spent notes and outputs.
+///
+/// The returned bundle will have no proof or signatures; these can be applied with
+/// [`Bundle::create_proof`] and [`Bundle::apply_signatures`] respectively.
+pub fn bundle<V: TryFrom<i64>>(
+    mut rng: impl RngCore,
+    mut spends: Vec<SpendInfo>,
+    mut outputs: Vec<OutputInfo>,
+    bundle_type: BundleType,
+) -> Result<Option<Bundle<InProgress<Unproven, Unauthorized>, V>>, BuildError> {
+    let (flags, anchor) = bundle_type.bundle_config();
+
+    let num_requested_spends = spends.len();
+    if !flags.spends_enabled() && num_requested_spends > 0 {
+        return Err(BuildError::SpendsDisabled);
+    }
 
-        // Determine the value balance for this bundle, ensuring it is valid.
-        let value_balance = pre_actions
-            .iter()
-            .fold(Some(ValueSum::zero()), |acc, action| {
-                acc? + action.value_sum()
-            })
-            .ok_or(OverflowError)?;
+    for spend in &spends {
+        if spend.note.value() != NoteValue::zero() {
+            let cm = spend.note.commitment();
+            let path_root = spend.merkle_path.root(cm.into());
+            if path_root != anchor {
+                return Err(BuildError::AnchorMismatch);
+            }
+        }
```

Comment:
Then this would become something like:
```suggestion
        spend.check_anchor_consistency(anchor)
            .map_err(|_| BuildError::AnchorMismatch)?;
```